Fix for #9463: iterator rewriter to report missing types gracefully

jcouv opened the pull request:
Two changes here:  

- Adding type checks before iterator rewrite starts
- Refactoring a large VB method in the iterator rewriter to align to C# implementation, which has a side-effect of re-ordering some methods that are generated

A few notes:

- I ended opting for reporting multiple diagnostics (not just the first one)
- I didn't end up storing the types between the place they are checked and the place they are used
- if this approach is acceptable, I will implement a similar change for async rewriter as next step
- some files had a unicode signature character at the start, which I removed

CC @dotnet/roslyn-compiler

AlekseyTs commented on src/Compilers/CSharp/Portable/Lowering/IteratorRewriter/IteratorRewriter.cs:
```diff
@@ -74,7 +75,91 @@ internal partial class IteratorRewriter : StateMachineRewriter
+        }
+
+        // returns true if any type we depend on is not present
+        protected bool EnsureAllSymbolsAndSignature()
```
Consider renaming to "VerifyPresenceOfRequiredAPIs" and changing to return true in case of success.

jcouv replied:
I agree. I kept the `false` return to be consistent with the VB implementation, but I'll change both.

AlekseyTs commented on src/Compilers/CSharp/Portable/Lowering/IteratorRewriter/IteratorRewriter.cs:
```diff
@@ -74,7 +75,91 @@ internal partial class IteratorRewriter : StateMachineRewriter
+
+            // IEnumerator<T>
+            EnsureSpecialPropertyAsMember(SpecialMember.System_Collections_Generic_IEnumerator_T__Current,
+                                            SpecialType.System_Collections_Generic_IEnumerator_T, ref hasErrors);
```
Is IEnumerator\<T\> always required? For example, in this case?  
```
    IEnumerable Test()
    {
        yield return null;
    }
```

jcouv replied:
Yes, the `GenerateEnumeratorImplementation` method always gets this type and uses it, and that method is always called when rewriting. 

As far as I can tell, that is also true on the VB side, but the VB `EnsureAllSymbolsAndSignature` has some conditional checks that didn't make sense to me. I'll stop by to discuss.

AlekseyTs commented on src/Compilers/CSharp/Portable/Lowering/IteratorRewriter/IteratorRewriter.cs:
```diff
@@ -74,7 +75,91 @@ internal partial class IteratorRewriter : StateMachineRewriter
+
+        private void EnsureSpecialProperty(SpecialMember type, ref bool hasErrors)
+        {
+            EnsureAvailableType(() => F.SpecialProperty(type).GetMethod, ref hasErrors);
```
What happens if GetMethod returns null?

AlekseyTs commented on src/Compilers/CSharp/Portable/Lowering/IteratorRewriter/IteratorRewriter.cs:
```diff
@@ -74,7 +75,91 @@ internal partial class IteratorRewriter : StateMachineRewriter
+            EnsureAvailableType(() =>
+            {
+                var sType = F.SpecialType(type).Construct(_elementType);
+                return F.SpecialProperty(property).GetMethod.AsMember(sType);
```
I don't see any good reason for calling AsMember for the purpose of checking existence of a property. Also what happens if GetMethod is null?

AlekseyTs commented on src/Compilers/CSharp/Portable/Lowering/IteratorRewriter/IteratorRewriter.cs:
```diff
@@ -74,7 +75,91 @@ internal partial class IteratorRewriter : StateMachineRewriter
+            }, ref hasErrors);
+        }
+
+        private void EnsureAvailableType(Func<Symbol> getType, ref bool hasErrors)
```
The name of the method is misleading, given how it is used. Also, I do not think we should be using lambdas and rely on BoundNodeFactory for the purpose of this checks. Let's talk offline in more details.

cston commented on src/Compilers/CSharp/Portable/Lowering/IteratorRewriter/IteratorRewriter.cs:
```diff
@@ -74,7 +75,91 @@ internal partial class IteratorRewriter : StateMachineRewriter
+            // IEnumerator
+            EnsureSpecialProperty(SpecialMember.System_Collections_IEnumerator__Current, ref hasErrors);
+            EnsureSpecialMethod(SpecialMember.System_Collections_IEnumerator__MoveNext, ref hasErrors);
+            EnsureSpecialMethod(SpecialMember.System_Collections_IEnumerator__Reset, ref hasErrors);
```
It looks like each `EnsureSpecial*` method calls `F.SpecialType(type)`. Consider getting the type once for each special type.

cston commented on src/Compilers/CSharp/Portable/Lowering/IteratorRewriter/IteratorRewriter.cs:
```diff
@@ -74,7 +75,91 @@ internal partial class IteratorRewriter : StateMachineRewriter
+        }
+
+        private void EnsureSpecialProperty(SpecialMember type, ref bool hasErrors)
+        {
```
Rename `type` to `property`.

cston commented on src/Compilers/CSharp/Portable/Lowering/IteratorRewriter/IteratorRewriter.cs:
```diff
@@ -74,7 +75,91 @@ internal partial class IteratorRewriter : StateMachineRewriter
+            EnsureAvailableType(() => F.SpecialProperty(type).GetMethod, ref hasErrors);
+        }
+
+        private void EnsureSpecialMethod(SpecialMember type, ref bool hasErrors)
```
Rename `type` to `method`.

jcouv commented:
@AlekseyTs @cston 
This revised PR should be much more code-review-friendly.

cston commented on src/Compilers/CSharp/Portable/Lowering/IteratorRewriter/IteratorRewriter.cs:
```diff
@@ -74,7 +75,61 @@ internal partial class IteratorRewriter : StateMachineRewriter
+            if (!Binder.TryGetSpecialTypeMember(F.Compilation, member, body.Syntax, diagnostics, out symbol))
+            {
+                hasErrors = true;
+            }
```
`hasErrors` is unchanged if the symbol has use-site errors. That's different from the behavior of `EnsureSpecialType` and the VB version of this method. Is that intentional? (Do we want to rewrite in that case?)

jcouv replied:
You are correct, that was not intentional. 
In the case of use-site errors, the diagnostics are created, but `hasErrors` isn't set and we continue with rewriting. But we shouldn't.
I will tighten this case and find a test to hit it.

AlekseyTs replied:
Consider making these methods static and passing a specially allocated DiagnosticBag to them. Use this bag for all checks, at the end you just need to check if there are errors in the bag and merge all diagnostics in. No need to have hasErrors parameter.

AlekseyTs commented on src/Compilers/CSharp/Portable/Lowering/IteratorRewriter/IteratorRewriter.cs:
```diff
@@ -74,7 +75,61 @@ internal partial class IteratorRewriter : StateMachineRewriter
+            return rewriter.Rewrite();
+        }
+
+        // returns true if all types we need are present
```
Use standard doc comment (///). Also this isn't just about types, it is also about members. It isn't just about presence, it is also about "goodness".

AlekseyTs commented on src/Compilers/CSharp/Portable/Lowering/IteratorRewriter/IteratorRewriter.cs:
```diff
@@ -74,7 +75,61 @@ internal partial class IteratorRewriter : StateMachineRewriter
+        private void EnsureSpecialType(SpecialType type, ref bool hasErrors)
+        {
+            NamedTypeSymbol symbol = Binder.GetSpecialType(F.Compilation, type, body.Syntax, diagnostics);
+            if (symbol.HasUseSiteError)
```
This check would be redundant if we switch to checking for presence of errors in the bag.

AlekseyTs commented on src/Compilers/CSharp/Portable/Lowering/IteratorRewriter/IteratorRewriter.cs:
```diff
@@ -74,7 +75,61 @@ internal partial class IteratorRewriter : StateMachineRewriter
+            return !hasErrors;
+        }
+
+        private void EnsureSpecialMember(SpecialMember member, ref bool hasErrors)
```
It feels like for properties we should also verify presence and goodness of getters.

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Lowering/IteratorRewriter/IteratorRewriter.vb:
```diff
@@ -79,6 +79,7 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
             Return rewriter.Rewrite()
         End Function
 
+        ' Returns true if there are errors
```
Use standard doc comment

AlekseyTs commented on src/Compilers/CSharp/Portable/Lowering/IteratorRewriter/IteratorRewriter.cs:
```diff
@@ -74,7 +75,61 @@ internal partial class IteratorRewriter : StateMachineRewriter
+            EnsureSpecialMember(SpecialMember.System_IDisposable__Dispose, ref hasErrors);
+
+            // IEnumerator
+            EnsureSpecialType(SpecialType.System_Collections_IEnumerator, ref hasErrors);
```
The same is applicable to other types in a similar situation.

AlekseyTs replied:
It might still make sense to do this if this improves diagnostics.

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Lowering/StateMachineRewriter/StateMachineRewriter.vb:
```diff
@@ -375,6 +375,13 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
         End Sub
 
+        Friend Sub EnsureSpecialMember(type As SpecialMember, <[In], Out> ByRef hasErrors As Boolean)
+            Dim sType = Me.F.SpecialMember(type)
```
The name of the local is misleading.

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Lowering/IteratorRewriter/IteratorRewriter.vb:
```diff
@@ -95,22 +96,25 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
-
-                EnsureSpecialType(SpecialType.System_Collections_IEnumerator, hasErrors)
+            ' IEnumerator
+            EnsureSpecialType(SpecialType.System_Collections_IEnumerator, hasErrors)
```
Here, however, the check for the type isn't redundant because EnsureSpecialMember doesn't check use-site error on the containing type for the purpose of setting hasErrors to true. The diagnostics will be reported though.

AlekseyTs commented on src/Compilers/VisualBasic/Test/Emit/CodeGen/CodeGenIterators.vb:
```diff
@@ -1688,5 +1689,139 @@ End Class
+"
+            Dim mslib = CreateCompilation({Parse(corlib)})
+            Dim mslibRef = mslib.EmitToImageReference()
+            Dim compilation = CreateCompilation({Parse(source)}, references:={mslibRef})
```
Can we create a single compilation with both sources instead?

AlekseyTs commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenIterators.cs:
```diff
@@ -2394,5 +2394,127 @@ public System.Collections.IEnumerable SomeNumbers()
+}";
+            var mslib = CreateCompilation(new[] { Parse(corlib) });
+            var mslibRef = mslib.EmitToImageReference();
+            var compilation = CreateCompilation(new[] { Parse(source) }, references: new[] { mslibRef });
```
Can we create a single compilation with both sources instead?

AlekseyTs commented on src/Compilers/VisualBasic/Test/Emit/CodeGen/CodeGenIterators.vb:
```diff
@@ -1688,5 +1689,139 @@ End Class
+            Dim mslibRef = mslib.EmitToImageReference()
+            Dim compilation = CreateCompilation({Parse(source)}, references:={mslibRef})
+
+            compilation.VerifyEmitDiagnostics(
```
Please use AssertEmitDiagnostics helper instead for VB.

AlekseyTs commented on src/Compilers/VisualBasic/Test/Emit/CodeGen/CodeGenIterators.vb:
```diff
@@ -1688,5 +1689,139 @@ End Class
+            Dim compilation = CreateCompilation({Parse(source)}, references:={mslibRef})
+
+            ' No error about IEnumerable
+            compilation.VerifyEmitDiagnostics(
```
Same comment as above.

jcouv commented:
@AlekseyTs @cston When you have some time, the new commit should address your previous feedback. Thanks

cston commented on src/Compilers/CSharp/Portable/Lowering/IteratorRewriter/IteratorRewriter.cs:
```diff
@@ -74,7 +75,84 @@ internal partial class IteratorRewriter : StateMachineRewriter
+                diagnostics.AddRange(bag);
+                return false;
+            }
+            return true;
```
```bag.Free()``` before returning.

cston commented on src/Compilers/CSharp/Portable/Lowering/IteratorRewriter/IteratorRewriter.cs:
```diff
@@ -74,7 +75,84 @@ internal partial class IteratorRewriter : StateMachineRewriter
+        private void EnsureSpecialPropertyGetter(SpecialMember member, DiagnosticBag bag)
+        {
+            PropertySymbol symbol = (PropertySymbol)EnsureSpecialMember(member, bag);
+            if (symbol != null)
```
Cast to object in comparison, here and for `getter` below.

cston replied:
Please use ```(object)symbol != null``` and ```(object)getter == null```.

AlekseyTs replied:
+1

jcouv replied:
Agreed, sorry I lost that one in the reverting of code.

cston commented on src/Compilers/CSharp/Portable/Lowering/IteratorRewriter/IteratorRewriter.cs:
```diff
@@ -74,7 +75,84 @@ internal partial class IteratorRewriter : StateMachineRewriter
+                if (getter.HasUseSiteError)
+                {
+                    bag.Add(new CSDiagnostic(getter.GetUseSiteDiagnostic(), body.Syntax.Location));
+                }
```
Does this handle an inaccessible getter? Consider using ```EnsureSpecialMember(IEnumerator__get_Current)``` in addition to ```EnsureSpecialMember(IEnumerator__Current)``` instead. Then `EnsureSpecialPropertyGetter` is not needed.

jcouv replied:
@cston, I'm trying to verify the existence of a member on an interface. When I tried adding accessibility modifiers in the test, I got an error: `error CS0275: accessibility modifiers may not be used on accessors in an interface`, so I think we're ok.
I will take your suggestion though and remove `EnsureSpecialPropertyGetter`. Thanks

AlekseyTs replied:
@jcouv I think this needs more thinking. Nothing ensures that IEnumerator__get_Current is a getter for IEnumerator__Current, that should be verified.

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Lowering/StateMachineRewriter/StateMachineRewriter.vb:
```diff
@@ -375,6 +375,10 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
             End If
         End Sub
 
+        Friend Function EnsureSpecialMember(member As SpecialMember, bag As DiagnosticBag) As Symbol
```
Consider adding similar overload for EnsureSpecialType to be consistent and rely on errors in the bag for all checks.

jcouv replied:
@AlekseyTs I'll do that, but I won't be able to convert all the checks to DiagnosticBag just yet.

See the first line:
```VB
        Friend Overrides Function EnsureAllSymbolsAndSignature() As Boolean
            Dim hasErrors As Boolean = MyBase.EnsureAllSymbolsAndSignature

            If Me.Method.IsSub OrElse Me._elementType.IsErrorType Then
                hasErrors = True
            End If

            Dim bag = DiagnosticBag.GetInstance()
```

AlekseyTs replied:
We can return from that ```If``` right away, no reason to continue.

jcouv commented:
@cston @AlekseyTs 
See latest commit for addressing your feedback.
From discussion with Aleksey and after checking the code, there are indeed slight differences in the validation of property getter in different places in the compiler. For now, I prefer to keep the check as close as possible to how the IteratorRewriter will get those getters, so I added the method to check properties back in.

jcouv commented:
@cston @AlekseyTs please review.

cston commented on src/Compilers/CSharp/Portable/Lowering/IteratorRewriter/IteratorRewriter.cs:
```diff
@@ -74,7 +75,91 @@ internal partial class IteratorRewriter : StateMachineRewriter
+            if (symbol != null)
+            {
+                var getter = symbol.GetMethod;
+                if (getter == null )
```
Extra space before close paren.

cston commented on src/Compilers/CSharp/Portable/Lowering/IteratorRewriter/IteratorRewriter.cs:
```diff
@@ -74,7 +75,91 @@ internal partial class IteratorRewriter : StateMachineRewriter
+            finally
+            {
+                bag.Free();
+            }
```
```try ... finally``` is typically avoided for freeing pooled instances. Instead consider something like the following here and in VB.
```
bool hasErrors = bag.HasAnyErrors();
if (hasErrors)
{
    diagnostics.AddRange(bag);
}
bag.Free();
return !hasErrors;
```

cston commented on src/Compilers/VisualBasic/Portable/Lowering/StateMachineRewriter/StateMachineRewriter.vb:
```diff
@@ -375,6 +379,31 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
+                Dim getter = symbol.GetMethod
+
+                If getter Is Nothing Then
+                    Binder.ReportDiagnostic(Diagnostics, Body.Syntax, ERRID.ERR_NoGetProperty1, CustomSymbolDisplayFormatter.QualifiedName(symbol))
```
`bag` rather than `Diagnostics`? Here and below.

jcouv replied:
There is already `Diagnostics` in scope. The purpose here is to capture diagnostics in a separate bag, so I picked a distinctive name.

AlekseyTs replied:
It is the wrong bag, it should go into parameter.

jcouv replied:
I understand now. Thanks.

AlekseyTs commented on src/Compilers/CSharp/Portable/Lowering/IteratorRewriter/IteratorRewriter.cs:
```diff
@@ -74,7 +75,91 @@ internal partial class IteratorRewriter : StateMachineRewriter
+                    Binder.Error(bag, ErrorCode.ERR_PropertyLacksGet, body.Syntax, symbol);
+                    return;
+                }
+                if (getter.HasUseSiteError)
```
Consider getting UseSiteDiagnostic and reporting it if it is not null regardless whether it is an error or not. Warnings are very uncommon, no reason to optimize for them. Also, consider adding an empty line between 'if' statements.

jcouv commented:
@cston @AlekseyTs Latest commit fixes the bug with wrong diagnostic bag, and addresses other suggestions.

cston commented on src/Compilers/VisualBasic/Portable/Lowering/IteratorRewriter/IteratorRewriter.vb:
```diff
@@ -79,39 +79,54 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
+                Return False
+            Finally
+                bag.Free()
+            End Try
```
Remove ```Try ... Finally``` for consistency with C#.

cston commented:
LGTM

AlekseyTs commented:
LGTM

jcouv commented:
@dotnet-bot retest prtest/win/dbg/unit32 please
// Previous failure: http://dotnet-ci.cloudapp.net/job/roslyn_prtest_win_dbg_unit32/5330/
// Retest reason: transient failure
